Bug Report: Return error if template name erroneous

### SUMMARY
 generated cli produces output that is correct but asserts wrongly with respect to expected cli 

### ISSUE TYPE
- Bug report fix

### COMPONENT NAME
nso_template.py

### NSO-Mock-Template VERSION
```
0.1.1
```

### SYSTEM SPECS
```
Mac OS Sierra
10.12.6
```

### ERROR EXAMPLE
```

================================================================================================= FAILURES =================================================================================================
__________________________________________________________________________________ test_vrf_ipv4_cci_gw_routing_template ___________________________________________________________________________________

    def test_vrf_ipv4_cci_gw_routing_template():
       "unittest to test if an ACL XML generates correct Cisco CLI"
       # Dummy that replicates ncs.application.Template behavior
       template = 'vrf_ipv4_cci_gw_routing_templatexy'
       service_variables = dict(
        cci_gw_hostname='myccigw',
        customer_vrf_name='springpath-1',
        customer='SPRINGPATH-1',
        ip_address_cci_gw1_tunnel='10.239.128.129',
        customer_asn='64704',
        crosslink_ip='173.36.246.134',
        cci_gw2='sjc12-cci-gw1',
        outside_vlan='1782',
        vrf_bridging_subnet='10.239.128.112',
        cci_gw1_tunnel_subnet='10.239.128.128',
        crosslink_subnet='173.36.246.132',
       )
    
       expected_cli = """router eigrp as109
           address-family ipv4 unicast vrf springpath-1 autonomous-system 1882
            !
            af-interface default
             passive-interface
            exit-af-interface
            !
            af-interface Port-channel1.1782
             no passive-interface
            !
            af-interface Port-channel15.1782
             no passive-interface
            exit-af-interface
            !
            topology base
             redistribute bgp 109 metric 10000000 100 255 1 1500 route-map springpath-1-vrf-subnets
            exit-af-topology
            network 10.239.128.112 0.0.0.15
            network 10.239.128.128 0.0.0.3
            network 173.36.246.132 0.0.0.3
            !
          router bgp 109
           address-family ipv4 vrf springpath-1
            redistribute eigrp 1882 metric 1 route-map SPRINGPATH-1-EIGRP-BGP
            neighbor 10.239.128.129 remote-as 64704
            neighbor 10.239.128.129 description springpath-1
            neighbor 10.239.128.129 activate
            neighbor 10.239.128.129 maximum-prefix 500
            neighbor 10.239.128.129 next-hop-self
            neighbor 10.239.128.129 route-map springpath-1-vrf-subnets in
            neighbor 173.36.246.134 remote-as 109
            neighbor 173.36.246.134 description sjc12-cci-gw1 iBGP
            !
          ip route vrf springpath-1 192.0.2.1 255.255.255.255 Null0"""
    
       generated_cli = main.apply_nso_template("service", template, service_variables)
       print generated_cli
>      assert expected_cli in generated_cli["native"]['devices'][0]["data"]
E      IndexError: list index out of range

test.py:101: IndexError
------------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------------
{'native': {'devices': []}}

```
### SUGGESTED FIX
```
this was a template name error and not an index error :) 
```

--- a/MonkeyTemplate/nso_template.py
+++ b/MonkeyTemplate/nso_template.py
@@ -92,6 +92,7 @@ class MonkeyTemplate(object):
 class MonkeyResult(object):
 
     def __init__(self, result):
+        self.raw = result
         self.result = self._parse_result(result)
 
     def _parse_result(self, result):

--- a/MonkeyTemplate/test.py
+++ b/MonkeyTemplate/test.py
@@ -28,22 +28,22 @@ ncs.template.Template = MonkeyTemplate # Monkey patch
 def test_make_ncs_list():
     monkey_template = MonkeyTemplate("service")
     service_variables = ncs.template.Variables()
-    service_variables.add('lab_gateway', "netsim")
-    assert monkey_template._make_ncs_list(service_variables) == "variable { name lab_gateway value netsim } "
+    service_variables.add('lab_gateway', "lab-gw0")
+    assert monkey_template._make_ncs_list(service_variables) == "variable { name lab_gateway value lab-gw0 } "
 
 def test_ncs_cli():
     monkey_template = MonkeyTemplate("service")
     service_variables = ncs.template.Variables()
-    service_variables.add('lab_gateway', "netsim")
-    assert monkey_template._make_ncs_cli("ipv6-acl", service_variables) == "ipv6-acl variable { name lab_gateway value netsim } "
-    assert monkey_template._open_ncs_cli("ipv6-acl variable { name lab_gateway value netsim } ")
+    service_variables.add('lab_gateway', "lab-gw0")
+    assert monkey_template._make_ncs_cli("ipv6-acl", service_variables) == "ipv6-acl variable { name lab_gateway value lab-gw0 } "
+    assert monkey_template._open_ncs_cli("ipv6-acl variable { name lab_gateway value lab-gw0 } ")
     cli_result = monkey_template._get_results()
     assert cli_result[0] == "Error: Python cb_action error. Unknown error (66): A variable value has not been assigned to: building_id, lab_id, lab_v6_p2p, lab_v6_prefix\n"
 
 def test_proper_cli():
     monkey_template = MonkeyTemplate("service")
     service_variables = ncs.template.Variables()
-    service_variables.add('lab_gateway', "netsim")
+    service_variables.add('lab_gateway', "lab-gw0")
     service_variables.add('lab_id', "12342")
     service_variables.add('building_id', "test1")
     service_variables.add('lab_v6_p2p', "2001::2")
@@ -63,7 +63,7 @@ def test_proper_cli():
 def test_dict_to_ncs_vars():
     # NSO Generic Template applicator helper tests
     variables ={'lab_id': "12342",
-                'lab_gateway': "netsim",
+                'lab_gateway': "lab-gw0",
                 'building_id': "test1",
                 'lab_v6_p2p': "2001::2",
                 'lab_v6_prefix': "2001::2/64",
@@ -72,7 +72,7 @@ def test_dict_to_ncs_vars():
 
 def test_apply_nso_template():
     variables ={'lab_id': "12342",
-                'lab_gateway': "netsim",
+                'lab_gateway': "lab-gw0",
                 'building_id': "test1",
                 'lab_v6_p2p': "2001::2",
                 'lab_v6_prefix': "2001::2/64",
@@ -181,3 +181,24 @@ native {
     assert "dummy2" in devices
     assert parsed["native"]['devices'][0]["data"] == "cli commands a b c\n            commands e f g"
     assert parsed["native"]['devices'][1]["data"] == "cli commands h i j\n            commands k l m"
+
+def test_MonkeyResult_raw():
+    """Test that the .raw field of the MonkeyResult.
+
+    """
+    result = """\
+native {
+        device {
+            name dummy
+            data cli commands a b c
+            commands e f g
+        }
+        device {
+            name dummy2
+            data cli commands h i j
+            commands k l m
+        }
+    }"""
+    mock_template =  MonkeyResult(result)
+
+    assert result == mock_template.raw

--- a/MonkeyTemplate/test.py
+++ b/MonkeyTemplate/test.py
@@ -28,22 +28,22 @@ ncs.template.Template = MonkeyTemplate # Monkey patch
 def test_make_ncs_list():
     monkey_template = MonkeyTemplate("service")
     service_variables = ncs.template.Variables()
-    service_variables.add('lab_gateway', "lab-gw0")
-    assert monkey_template._make_ncs_list(service_variables) == "variable { name lab_gateway value lab-gw0 } "
+    service_variables.add('lab_gateway', "netsim")
+    assert monkey_template._make_ncs_list(service_variables) == "variable { name lab_gateway value netsim } "
 
 def test_ncs_cli():
     monkey_template = MonkeyTemplate("service")
     service_variables = ncs.template.Variables()
-    service_variables.add('lab_gateway', "lab-gw0")
-    assert monkey_template._make_ncs_cli("ipv6-acl", service_variables) == "ipv6-acl variable { name lab_gateway value lab-gw0 } "
-    assert monkey_template._open_ncs_cli("ipv6-acl variable { name lab_gateway value lab-gw0 } ")
+    service_variables.add('lab_gateway', "netsim")
+    assert monkey_template._make_ncs_cli("ipv6-acl", service_variables) == "ipv6-acl variable { name lab_gateway value netsim } "
+    assert monkey_template._open_ncs_cli("ipv6-acl variable { name lab_gateway value netsim } ")
     cli_result = monkey_template._get_results()
     assert cli_result[0] == "Error: Python cb_action error. Unknown error (66): A variable value has not been assigned to: building_id, lab_id, lab_v6_p2p, lab_v6_prefix\n"
 
 def test_proper_cli():
     monkey_template = MonkeyTemplate("service")
     service_variables = ncs.template.Variables()
-    service_variables.add('lab_gateway', "lab-gw0")
+    service_variables.add('lab_gateway', "netsim")
     service_variables.add('lab_id', "12342")
     service_variables.add('building_id', "test1")
     service_variables.add('lab_v6_p2p', "2001::2")
@@ -63,7 +63,7 @@ def test_proper_cli():
 def test_dict_to_ncs_vars():
     # NSO Generic Template applicator helper tests
     variables ={'lab_id': "12342",
-                'lab_gateway': "lab-gw0",
+                'lab_gateway': "netsim",
                 'building_id': "test1",
                 'lab_v6_p2p': "2001::2",
                 'lab_v6_prefix': "2001::2/64",
@@ -72,7 +72,7 @@ def test_dict_to_ncs_vars():
 
 def test_apply_nso_template():
     variables ={'lab_id': "12342",
-                'lab_gateway': "lab-gw0",
+                'lab_gateway': "netsim",
                 'building_id': "test1",
                 'lab_v6_p2p': "2001::2",
                 'lab_v6_prefix': "2001::2/64",

--- a/README.md
+++ b/README.md
@@ -115,6 +115,18 @@ With the following XML template:
 
 ## Common issues
 
+### template.apply() returns `{'native':{'devices':[]}}`
+
+This is an issue where there is something that went wrong in the application of the template.
+
+To see the raw result from NSO print `template.apply().raw` this will provide the most info
+
+Common reasons:
+  1. There were 'No Modifications to Commit' -> ie the netsim already has the config
+  2. A variable was failed to be applied
+  3. The device name is mis-typed / nonexistent
+  4. NED errors applying the template
+
 ### template.apply() mock functionality returns `syntax error: unknown command`
 
 It indicates that the apply-config-template needs to be installed inside NSO.

